provider/haproxy/haproxy.go

```diff
@@ -160,6 +199,236 @@ func (lbp *Provider) GetPublicEndpoints(configName string) []string {
+					}
+					if err == nil && process == nil {
+						return true
+					}
+					logrus.Infof("IsEndpointDrained: old haproxy process found pid %v, name %v", pid, process.Executable())
+				}
+			} else {
+				//check if current pid stats are scur = 0 AND no other pid exists.
+				//read stats
+				scur, _, err := lbp.drainMgr.readCurrentStats(ep)
+				if err != nil {
+					logrus.Errorf("IsEndpointDrained: Error %v", err)
+					return false
+				}
+				logrus.Debugf("IsEndpointDrained: scur %v Endpoint %v", scur, ep.Name)
+				oldPidExists, err := lbp.drainMgr.olderPidExists()
+				if err != nil {
+					logrus.Errorf("AddEndpointForDrain: Error %v listing older haproxy processes, drain till timeout %v", err, ep.DrainTimeout)
+					return false
+				}
+				if !oldPidExists && scur == "0" {
+					return true
+				}
+			}
+		} else {
+			//read stats
+			scur, _, err := lbp.drainMgr.readCurrentStats(ep)
+			if err != nil {
+				logrus.Errorf("IsEndpointDrained: Error %v", err)
+				return false
+			}
+			logrus.Debugf("IsEndpointDrained: scur %v Endpoint %v", scur, ep.Name)
+			if scur == "0" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (dm *drainMgr) readCurrentStats(ep *config.Endpoint) (string, string, error) {
+	//read stats
+	stats, err := dm.ReadStats()
+	if err != nil {
+		return "", "", fmt.Errorf("Error %v reading haproxy stats on  %v", err, dm.haproxySock)
+	}
+	if endpointStat, ok := stats[ep.Name]; ok {
+		return endpointStat["scur"], endpointStat["pxname"], nil
+	}
+	return "", "", fmt.Errorf("Error %v finding endpoint %v in haproxy stats", err, ep.Name)
+}
+
+func (dm *drainMgr) olderPidExists() (bool, error) {
+	output, err := exec.Command("pidof", "haproxy").Output()
+	if err != nil {
+		return true, fmt.Errorf("error listing old process %v", err)
+	}
+	processes := string(output)
+	if string(output) != "" {
+		logrus.Debugf("Output of old process %v", processes)
+	}
+
+	parts := strings.Split(processes, " ")
+	if len(parts) > 1 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (dm *drainMgr) AddEndpointForDrain(ep *config.Endpoint, pid string) bool {
+	//check if drain is needed by looking at Stats
+	if ep != nil && ep.Name != "" {
+		scur, pxname, err := dm.readCurrentStats(ep)
+		if err != nil {
+			logrus.Errorf("AddEndpointForDrain: Error %v", err)
+			return false
+		}
+		logrus.Debugf("AddEndpointForDrain: scur %v Endpoint %v", scur, ep.Name)
+		if scur == "" || scur == "0" {
+			//stats are zero on the current pid, check if any older pid around.
+			oldPidExists, err := dm.olderPidExists()
+			if err != nil {
+				logrus.Errorf("AddEndpointForDrain: Error %v listing older haproxy processes, drain till timeout %v", err, ep.DrainTimeout)
+			}
+			logrus.Debugf("AddEndpointForDrain: pid %v oldpid %v", pid, oldPidExists)
+
+			if !oldPidExists {
+				logrus.Debugf("AddEndpointForDrain: Drain not needed for Endpoint %v", ep.Name)
+				return false
```

Comment:
So, if scur is already at zero and no older pids exist, we basically say it is good to go. Is that right? I see that in this scenario we would not set the weight to zero.

Could that result in a new connection coming in?

Reply:
Yes good point, we can have a little window where in a new connection gets accepted. I say little window because soon enough the metadata change should set the weight to zero via config for this stopping container. 

But I will explicitly set the weight to zero on socket to be sure.